state/postgresql/v2/postgresql.go

```diff
@@ -63,36 +71,57 @@ func NewPostgreSQLStateStore(logger logger.Logger) state.Store {
 	return NewPostgreSQLStateStoreWithOptions(logger, Options{})
 }
 
+// NewPostgreSQLStateStoreWithTrueOptions creates a new instance of PostgreSQL state store v2 with Azure AD authentication and AWS IAM authentication disabled.
+// The v2 of the component uses a different format for storing data, always in a BYTEA column, which is more efficient than the JSONB column used in v1.
+// Additionally, v2 uses random UUIDs for etags instead of the xmin column, expanding support to all Postgres-compatible databases such as CockroachDB, etc.
+func NewPostgreSQLStateStoreWithTrueOptions(logger logger.Logger) state.Store {
+	return NewPostgreSQLStateStoreWithOptions(logger, Options{NoAWSIAM: true, NoAzureAD: true})
+}
+
 // NewPostgreSQLStateStoreWithOptions creates a new instance of PostgreSQL state store with options.
 func NewPostgreSQLStateStoreWithOptions(logger logger.Logger, opts Options) state.Store {
 	s := &PostgreSQL{
 		logger:        logger,
 		enableAzureAD: !opts.NoAzureAD,
+		enableAWSIAM:  !opts.NoAWSIAM,
 	}
 	s.BulkStore = state.NewDefaultBulkStore(s)
 	return s
 }
 
 // Init sets up Postgres connection and performs migrations
 func (p *PostgreSQL) Init(ctx context.Context, meta state.Metadata) error {
-	err := p.metadata.InitWithMetadata(meta, p.enableAzureAD)
+	var (
+		useAWS   bool
+		useAzure bool
+		err      error
+	)
+
+	awsIam, _ := metadata.GetMetadataProperty(meta.Properties, "UseAWSIAM")
+	useAWS = utils.IsTruthy(awsIam)
+	azureAd, _ := metadata.GetMetadataProperty(meta.Properties, "UseAzureAD")
+	useAzure = utils.IsTruthy(azureAd)
+
+	opts := pgauth.InitWithMetadataOpts{
+		AzureADEnabled: useAzure,
+		AWSIAMEnabled:  useAWS,
+	}
+
+	err = p.metadata.InitWithMetadata(meta, opts)
 	if err != nil {
-		p.logger.Errorf("Failed to parse metadata: %v", err)
 		return err
 	}
 
-	config, err := p.metadata.GetPgxPoolConfig()
+	config, err := p.metadata.GetPgxPoolConfig(ctx)
 	if err != nil {
-		p.logger.Error(err)
 		return err
 	}
 
 	connCtx, connCancel := context.WithTimeout(ctx, p.metadata.Timeout)
 	p.db, err = pgxpool.NewWithConfig(connCtx, config)
-	connCancel()
+	defer connCancel()
```

Comment:
@ItalyPaleAle can you confirm whether this should be deferred here? I know sometimes we want to immediately cancel the context.

Reply:
There is a ping later on in the func so yes I think deferring is fine